### PR TITLE
Allow for broadcasting of distribution parameters

### PR DESCRIPTION
**Pull request by neerajprad**
This ensures consistent broadcasting of parameters passed to distribution classes:
 - all scalars are upcast to Tensors / Variables. 
 - all Tensors / Variable arguments are broadcasted to the inferred shape in the constructor.
 - the Bernoulli distribution also supports scalar parameters now to keep it consistent with Gamma and Normal.

Note that this is needed for consistent handling of batch and event shapes [probtorch#30](https://github.com/probtorch/pytorch/issues/30). 

**Testing:** Unit test cases added for asserting on valid / invalid examples of parameter broadcasting.

**Issue:** Refer to [probtorch#32](https://github.com/probtorch/pytorch/issues/32) for more details. [Distribution design doc](https://docs.google.com/document/d/1wnABg0cdyaVMr-Xqz_brHBwnPJuzTeSk3Oqko2HHSfE/edit?usp=sharing)

**Note to reviewers** - This change is not backwards compatible. e.g. 
```
Normal(torch.Tensor([[0, 0]]), torch.Tensor([[1], [1]])).sample()
earlier -> Tensor of shape (1, 2); same as mean.size()
now -> Tensor of shape (2, 2); with full broadcasting
```
Also, let us get outstanding PRs, namely #3886 and #4129 merged first, so as to avoid creating further merge conflicts.

### CODE REVIEW

**Comment by neerajprad** [comment]
cc. @apaszke, @fritzo, @alicanb - Small PR which is ready for review, but let us wait until #3886 and #4129 are merged, before merging this.

File: torch/distributions/gamma.py
Comment (apaszke) [inline torch/distributions/gamma.py]: The part with upcasting alpha and beta to tensors should be part of a helper as well (it is repeated in one of the other PRs)

File: torch/distributions/utils.py
Comment (apaszke) [inline torch/distributions/utils.py]: Isn't this just `torch._C._infer_size(t1.size(), t2.size())`?

Comment (neerajprad) [reply torch/distributions/utils.py]: I didn't realize there was an already accessible helper utility; will use that instead.

File: torch/distributions/gamma.py
Comment (neerajprad) [reply torch/distributions/gamma.py]: I was thinking of addressing that separately, but it is closely related to this; so let me fix the way we are handling scalars too (that should fix the tests).

File: test/test_distributions.py
Comment (fritzo) [inline test/test_distributions.py]: nit: Gamma parameters must be positive. We don't currently check that, but it would be nice to make this test resilient to change in case we do add a check.

File: torch/distributions/normal.py
Comment (fritzo) [inline torch/distributions/normal.py]: If we want to support numbers then we'll also have to wrap them in Tensors or Variables. As @apaszke has mentioned, it would be nice to have a helper `broadcast_all(*params)` that could correctly deal with mismatched type (`float` vs `Tensor`) and shape (`(1,4)` vs `(5,1)`).

Comment (neerajprad) [reply torch/distributions/normal.py]: Yeah, I had not realized that numbers are also supported. Will add a helper utility to take care of this.

File: test/test_distributions.py
Comment (neerajprad) [reply test/test_distributions.py]: Will change. Thanks for pointing out.

File: torch/distributions/utils.py
Comment (fritzo) [inline torch/distributions/utils.py]: Hmm, I think this doesn't handle `Variable` case since `Variable` is lacking the `.new` method.

File: torch/distributions/utils.py
Comment (fritzo) [inline torch/distributions/utils.py]: Wow this turns out to be tricky. I wonder if it's simpler to special-case Variable vs Tensor vs Number?
```py
# Promote numbers to Variables or Tensors.
if any(isinstance(v, Variable) for v in values):
    template = next(v.data for v in values if isinstance(v, Variable))
    values = [Variable(template.new([v])) if isinstance(v, Number) else v for v in values]
elif any(is_tensor(v) for v in values):
    template = next(v for v in values if is_tensor(v))
    values = [template.new([v]) if isinstance(v, Number) else v for v in values]
else:  # Everything is a number.
    values = [Tensor([v]) for v in values]

# Broadcast all shapes. 
shape = _broadcast_shape([v.size() for v in values])
return [v.expand(shape) for v in values]
```
I believe it also saves memory to use `template.new([v])` rather than `template.new(shape).fill_(v)`.

File: torch/distributions/utils.py
Comment (neerajprad) [reply torch/distributions/utils.py]: But, `Variable.new` seems to give expected results. Does that not work?

```
In [5]: v.new()
Out[5]: Variable containing:[torch.FloatTensor with no dimension]
```

Comment (fritzo) [reply torch/distributions/utils.py]: Sorry, I must have tested on an older PyTorch. I'll delete this comment.

File: torch/distributions/utils.py
Comment (neerajprad) [reply torch/distributions/utils.py]: Yes, will make the optimization, and let me see if I can simplify this further.

File: torch/distributions/utils.py
Comment (alicanb) [reply torch/distributions/utils.py]: You need to add shape as input

Comment (alicanb) [reply torch/distributions/utils.py]: From what I understand, `.new` works as `torch.Tensor`, which returns 0d tensor if no arguments are given.

File: torch/distributions/utils.py
Comment (neerajprad) [reply torch/distributions/utils.py]: Removed the list comprehensions and the zipping logic in favor of two cases, and am just using simple for loops now, since we don't really need to create new lists and it is just easier to operate by tracking scalars / tensors via indices.

File: torch/distributions/utils.py
Comment (neerajprad) [reply torch/distributions/utils.py]: Yes, you are right. I didn't supply any arguments to new above. It should work the same for Variables too.

**Review by apaszke — APPROVED** [review]

**Pull request merged**
